### PR TITLE
save the WriteId + PartitionId

### PR DESCRIPTION
**Pull request by Alek5andr-Kotov**
- для пары WriteId, PartitionId создаётся служебная партиция
- идентификатор партиции сохраняется между рестартами таблетки PQ
- служебная партиция восстанавливается на старте таблетки PQ
- таблетка PQ подписывается на WriteId в LongTxService

### CODE REVIEW

File: ydb/core/persqueue/pq_impl.cpp [hunk 1]
```diff
@@ -1176,22 +1283,22 @@ void TPersQueue::AggregateAndSendLabeledCountersFor(const TString& group, const
 
 void TPersQueue::Handle(TEvPQ::TEvPartitionLabeledCounters::TPtr& ev, const TActorContext& ctx)
 {
-    auto it = Partitions.find(ev->Get()->Partition);
-    Y_ABORT_UNLESS(it != Partitions.end());
+    const auto partitionId = ev->Get()->Partition;
+    auto& partition = GetPartitionInfo(partitionId);
     const TString& group = ev->Get()->LabeledCounters.GetGroup();
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:1288]: Do not public labeled counters for shadow partitions. return right now.

File: ydb/core/persqueue/pq_impl.cpp [hunk 2]
```diff
@@ -1176,22 +1283,22 @@ void TPersQueue::AggregateAndSendLabeledCountersFor(const TString& group, const
...
 //    AggregateAndSendLabeledCountersFor(group, ctx);
 }
 
 void TPersQueue::Handle(TEvPQ::TEvPartitionLabeledCountersDrop::TPtr& ev, const TActorContext& ctx)
 {
-    auto it = Partitions.find(ev->Get()->Partition);
-    Y_ABORT_UNLESS(it != Partitions.end());
+    const auto partitionId = ev->Get()->Partition;
+    auto& partition = GetPartitionInfo(partitionId);
     const TString& group = ev->Get()->Group;
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:1299]: The same situation. Do not process labeled counters from shadow partitions.

File: ydb/core/persqueue/pq_impl.cpp [hunk 3]
```diff
@@ -2345,6 +2464,94 @@ void TPersQueue::HandleSplitMessageGroupRequest(ui64 responseCookie, const TActo
...
+                                                             const NKikimrClient::TPersQueuePartitionRequest& req,
+                                                             const TActorId& sender,
+                                                             const TActorContext& ctx)
+{
+    Y_ABORT_UNLESS(req.HasWriteId());
+
+    ui64 writeId = req.GetWriteId();
+    ui32 partitionId = req.GetPartition();
+
+    Y_ABORT_UNLESS(TxWrites.contains(writeId) && TxWrites[writeId].Partitions.contains(partitionId));
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:2507]: TODO: use the same logic in all 3 handlers. Create partition if needed, store event in queue. And process it only when shadow partition initied.

File: ydb/core/persqueue/pq_impl.cpp [hunk 4]
```diff
@@ -2345,6 +2464,94 @@ void TPersQueue::HandleSplitMessageGroupRequest(ui64 responseCookie, const TActo
...
+        if (partition.InitDone) {
+            const TActorId& actorId = partition.Actor;
+            TActorId pipeClient = ActorIdFromProto(req.GetPipeClient());
+
+            HandleGetOwnershipRequest(responseCookie, actorId, req, ctx, pipeClient, sender);
+        } else {
+            partition.GetOwnershipRequests.emplace_back(responseCookie, req, sender);
+        }
+    } else {
+        GetOwnershipRequests.push_back({responseCookie, req, sender});
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:2491]: Fill map TWrites/ShadowPartitionsId here. Create shadow partition actor only when map writeid/partitionid->shadow partition id is stored. In this case GetOwnershipRequests queue will be needed only in partition.

File: ydb/core/persqueue/pq_impl.cpp [hunk 5]
```diff
@@ -2345,6 +2464,94 @@ void TPersQueue::HandleSplitMessageGroupRequest(ui64 responseCookie, const TActo
...
+    Y_ABORT_UNLESS(ShadowPartitions.contains(shadowPartitionId));
+    const TPartitionInfo& partition = ShadowPartitions.at(shadowPartitionId);
+    Y_ABORT_UNLESS(partition.InitDone);
+    const TActorId& actorId = partition.Actor;
+    TActorId pipeClient = ActorIdFromProto(req.GetPipeClient());
+
+    HandleReserveBytesRequest(responseCookie, actorId, req, ctx, pipeClient, sender);
+}
+
+void TPersQueue::HandleWriteRequestForShadowPartition(const ui64 responseCookie,
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:2519]: Shadow->Supportive

File: ydb/core/persqueue/pq_impl.cpp [hunk 6]
```diff
@@ -2345,6 +2464,94 @@ void TPersQueue::HandleSplitMessageGroupRequest(ui64 responseCookie, const TActo
...
+                                       const TActorContext& ctx)
+{
+    if (req.HasCmdGetOwnership()) {
+        HandleGetOwnershipRequestForShadowPartition(responseCookie, req, sender, ctx);
+    } else if (req.HasCmdReserveBytes()) {
+        HandleReserveBytesRequestForShadowPartition(responseCookie, req, sender, ctx);
+    } else if (req.CmdWriteSize()) {
+        HandleWriteRequestForShadowPartition(responseCookie, req, ctx);
+    } else {
+        Y_ABORT("CmdGetOwnership, CmdReserveBytes or CmdWrite expected");
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:2551]: reply with error, not abort!

File: ydb/core/persqueue/pq_impl.cpp [hunk 7]
```diff
@@ -3031,6 +3309,7 @@ void TPersQueue::EndWriteTxs(const NKikimrClient::TResponse& resp,
 
     SendReplies(ctx);
     CheckChangedTxStates(ctx);
+    ForwardGetOwnershipToShadowPartitions(ctx);
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:3312]: Change this func to CreateSupportivePartitions

File: ydb/core/persqueue/pq_impl.cpp [hunk 8]
```diff
@@ -2981,11 +3194,72 @@ bool TPersQueue::CanProcessDeleteTxs() const
...
+        ui32 shadowPartitionId = txWrite.Partitions[partitionId];
+        Y_ABORT_UNLESS(ShadowPartitions.contains(shadowPartitionId));
+
+        CreateShadowPartitionActor(shadowPartitionId, ctx);
+
+        TPartitionInfo& partition = ShadowPartitions.at(shadowPartitionId);
+        partition.GetOwnershipRequests.emplace_back(params.Cookie, params.Request, params.Sender);
+
+        if (txWrite.LongTxSubscriptionStatus == NKikimrLongTxService::TEvLockStatus::STATUS_UNSPECIFIED) {
+            SubscribeWriteId(writeId, ctx);
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:3220]: Subscribe to write id when got new writeid, not when in persisted

File: ydb/core/persqueue/pq_impl.cpp [hunk 9]
```diff
@@ -3915,6 +4223,16 @@ void TPersQueue::ProcessSourceIdRequests(ui32 partitionId) {
     }
 }
 
+void TPersQueue::Handle(NLongTxService::TEvLongTxService::TEvLockStatus::TPtr& ev, const TActorContext& /*ctx*/)
+{
+    auto& record = ev->Get()->Record;
+    ui64 writeId = record.GetLockId();
+    if (TxWrites.contains(writeId)) {
```
Comment (alexnick88) [inline ydb/core/persqueue/pq_impl.cpp:4230]: If  lockStatus == NO TX and Writeid is presented in memory state  and no propose in memory state - delete it and all partitions

Comment (alexnick88) [reply ydb/core/persqueue/pq_impl.cpp:4230]: Change deletion of writeID to :
1) Move write id to  inmemory map to_be_deleted_write_id -> shadow_partitions
2) Send to all shadow partitions event ClearAllAndDie, wait for results
3) when all results for partitions in writeId received, shedule deletion of all info for write id in tablet state.

File: ydb/core/persqueue/pq_impl.cpp [hunk 10]
```diff
@@ -830,6 +832,101 @@ void TPersQueue::ReadTxInfo(const NKikimrClient::TKeyValueResponse::TReadResult&
     LOG_DEBUG_S(ctx, NKikimrServices::PERSQUEUE, "Tablet " << TabletID() << " LastStep " << LastStep << " LastTxId " << LastTxId);
 }
 
+void TPersQueue::InitPlanStep(const NKikimrPQ::TTabletTxInfo& info)
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:835]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/102de710b7d7f86c369fc839a8d009d44e69a7bc

File: ydb/core/persqueue/pq_impl.cpp [hunk 11]
```diff
@@ -830,6 +832,101 @@ void TPersQueue::ReadTxInfo(const NKikimrClient::TKeyValueResponse::TReadResult&
...
+
+    TPartitionInfo& partition = ShadowPartitions.at(shadowPartitionId);
+    partition.Actor = ctx.Register(CreatePartitionActor(shadowPartitionId,
+                                                        TopicConverter,
+                                                        Config,
+                                                        true,
+                                                        ctx));
+}
+
+void TPersQueue::InitTxWrites(const NKikimrPQ::TTabletTxWrites& info,
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:896]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/102de710b7d7f86c369fc839a8d009d44e69a7bc

File: ydb/core/persqueue/pq_impl.cpp [hunk 12]
```diff
@@ -830,6 +832,101 @@ void TPersQueue::ReadTxInfo(const NKikimrClient::TKeyValueResponse::TReadResult&
...
+        ctx.Send(ctx.SelfID, new TEvents::TEvPoisonPill());
+        return;
+    }
+
+    switch (read.GetStatus()) {
+    case NKikimrProto::OK: {
+        LOG_INFO_S(ctx, NKikimrServices::PERSQUEUE, "Tablet " << TabletID() << " has a tx writes info");
+
+        NKikimrPQ::TTabletTxWrites info;
+        Y_ABORT_UNLESS(info.ParseFromString(read.GetValue()));
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:862]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/fdb95e35cf6232ff1c4dee904a9d92d70cb9ff13

File: ydb/core/persqueue/pq_impl.cpp [hunk 13]
```diff
@@ -830,6 +832,101 @@ void TPersQueue::ReadTxInfo(const NKikimrClient::TKeyValueResponse::TReadResult&
...
+    }
+    case NKikimrProto::NODATA: {
+        LOG_INFO_S(ctx, NKikimrServices::PERSQUEUE, "Tablet " << TabletID() << " doesn't have tx writes info");
+
+        InitTxWrites();
+
+        break;
+    }
+    default:
+        Y_ABORT("Unexpected tx writes info read status: %d", read.GetStatus());
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/fdb95e35cf6232ff1c4dee904a9d92d70cb9ff13

File: ydb/core/protos/pqconfig.proto [hunk 14]
```diff
@@ -1009,9 +1009,17 @@ message TTransaction {
...
+    optional uint64 WriteId = 15;
 };
 
 message TTabletTxInfo {
     optional uint64 LastStep = 2;
     optional uint64 LastTxId = 3;
 }
+
+message TTabletTxWrites {
+    repeated uint64 WriteIds = 1;
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/protos/pqconfig.proto]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/918122865b0df0acaa683b9a0c89fd0932ea92f5

File: ydb/core/persqueue/pq_impl.cpp [hunk 15]
```diff
@@ -1108,12 +1206,21 @@ void TPersQueue::Handle(TEvPQ::TEvMetering::TPtr& ev, const TActorContext&)
     MeteringSink.IncreaseQuantity(ev->Get()->Type, ev->Get()->Quantity);
 }
 
+TPartitionInfo& TPersQueue::GetPartitionInfo(ui32 partitionId)
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:1204]: PR
https://github.com/ydb-platform/ydb/pull/1472

File: ydb/core/persqueue/pq_impl.cpp (hunk 5 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/4250769b8b3e3663d606bf739dcc74aa2542f078

File: ydb/core/persqueue/pq_impl.cpp (hunk 8 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/74c832f44de40a1c11abdeb95e1b95dd7266a9bc

File: ydb/core/persqueue/pq_impl.cpp (hunk 4 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/74c832f44de40a1c11abdeb95e1b95dd7266a9bc

File: ydb/core/persqueue/pq_impl.cpp (hunk 7 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/74c832f44de40a1c11abdeb95e1b95dd7266a9bc

File: ydb/core/persqueue/pq_impl.cpp (hunk 2 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:1297]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/5090d78400d2f4c82c3f358033a1b8b725aea96b

File: ydb/core/persqueue/pq_impl.cpp (hunk 1 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:1283]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/5090d78400d2f4c82c3f358033a1b8b725aea96b

File: ydb/core/persqueue/pq_impl.cpp (hunk 9 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp:4222]: We agreed to do it in a separate task. There I described the above algorithm
LOGBROKER-8871

File: ydb/core/persqueue/pq_impl.cpp (hunk 3 above)
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/d2b1192e2d969c93ca2b18e033bcbd38425aeecd

File: ydb/core/persqueue/pq_impl.cpp [hunk 16]
```diff
@@ -1210,17 +1317,29 @@ void TPersQueue::Handle(TEvPQ::TEvTabletCacheCounters::TPtr& ev, const TActorCon
         << "Counters. CacheSize " << CacheCounters.CacheSizeBytes << " CachedBlobs " << CacheCounters.CacheSizeBlobs);
 }
 
+bool TPersQueue::AllPartitionsInited() const
+{
+    return PartitionsInited == (Partitions.size() + ShadowPartitions.size());
```
[the remark this answers is not in the document]
Comment (Alek5andr-Kotov) [reply ydb/core/persqueue/pq_impl.cpp]: done
https://github.com/ydb-platform/ydb/pull/1037/commits/048869b1ef706dcc5343fa84b1d1e386bcada3e4

**Review by alexnick88 — DISMISSED** [review]

**Review by FloatingCrowbar — APPROVED** [review]

**Pull request merged**